### DISCUSSION TITLE
Using AbaqusUMATStress with generalized plane strain

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#20867] [Q&A Modules: Solid mechanics] [answered]
This combo does not work as-is. Not sure if by design or just an oversight.

In poking around in the code for the generalized plane strain problem, it looks like the `GeneralizedPlaneStrainUserObject` requires an entry from the elasticity tensor to compute the Jacobian contribution for the out-of-plane scalar variable ([see here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUserObject.C#L125-L130)). But the problem is that when using `AbaqusUMATStress`, a global elasticity tensor material does not always need to be built, and as such, MOOSE cannot access specific components of the tensor. The AbaqusUMATStress material does access the components of the tensor in UMAT-notation in order to contribute to the Jacobian for the displacement variables (I believe, [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C#L308-L326)), but that information does not make it over to the `GeneralizedPlaneStrainUserObject`.

So it seems like the information that it needs to do generalized plane strain is available from the UMAT interface, I'm just not sure if there's an "easy" way to get at it from the `GeneralizedPlaneStrainUserObject`.

**Comment by dschwen** [comment, marked as answer]
It was not in the statement of work. As far as I can see the way forward would involve a new `GeneralizedUMATPlainStrain` object that calls the UMAT to obtain both stress and the Jacobian component. The elasticity tensor is "secret" and might not even exist as such in a general UMAT.

**Reply by jessecarterMOOSE** [reply]
@dschwen Yeah it seems maybe we should have asked you for more :)

Anyway, just so I understand things, the `AbaqusUMATMaterial` [declares](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C#L63) a `Jacobian_mult` property, which gets consumed by [StressDivergence](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/kernels/StressDivergenceTensors.C#L69) and [GeneralizedPlaneStrainOffDiag](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C#L55) kernels (and probably more). Does it make sense for `GeneralizedPlaneStrainUserObject` to also use this property? Right now it uses the elasticity tensor but there are a number of occasions ([here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C#L64) and [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/ComputeLinearElasticStress.C#L48)) where the `Jacobian_mult` property is set equal to the elasticity tensor.

Guess I'm just looking for the easy way out. Besides, the `GeneralizedPlaneStrainUserObject` [only needs a *single* entry](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUserObject.C#L126-L130) from the elasticity tensor to compute the Jacobian. Seems a bit overkill to create a new class just to get at a single value. I could be way off base here, so please let me know if I'm misunderstanding things.
